Commit authors in changelog

TwistingTwists opened the pull request:
fix: #73 

### Contributor checklist
- [x] My commit messages follow the [Conventional Commit Message Format](https://gist.github.com/stephenparish/9941e89d80e2bc58a153#format-of-the-commit-message)
      For example: `fix: Multiply by appropriate coefficient`, or
      `feat(Calculator): Correctly preserve history`
      Any explanation or long form information in your commit message should be
      in a separate paragraph, separated by a blank line from the primary message
- [ ] Bug fixes include regression tests
- [x] Features include unit/acceptance tests

TwistingTwists committed:
feat: git authors in the changelog

zachdaniel commented:
I think what we might want to do is add a configuration here that tells us to use the GitHub API to look up handles for each commit so we can properly build the `@`, and then if that configuration isn't set we include the author directly instead of computing a fake handle for them. Thoughts?

TwistingTwists commented:
Ha. That should be doable. 

https://stackoverflow.com/questions/44888187/get-github-username-through-primary-email

this only works if their email is publicly visible. it is doable. 
https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#search-users

TwistingTwists commented:
I don't assume people will hit rate limits with this API. but if they run this in CI, they might.

TwistingTwists commented:
> set we include the author directly

you mean, directly mention just the `git user.name` of the user, right?

TwistingTwists committed:
feat: lookup user by email on github to find correct @ names

zachdaniel commented:
Yeah, instead of deriving like an `@username` that may not exist.

TwistingTwists committed:
feat: use github_api to lookup username

- since usernames are looked up per commit, cache is implemented via
  persistent_term

TwistingTwists committed:
docs: update readme for github_handle_lookup?

TwistingTwists commented:
Made changes. 
1. look by git user.email via github api 
2. fallback to name (without `@`) 
3. added caching using persistent term.

TwistingTwists committed:
chore: mix format

zachdaniel commented on lib/git_ops/github.ex:
```diff
@@ -0,0 +1,90 @@
+  def find_user_by_email(email) do
+    cache_key = @cache_key_prefix <> email
+
+    case :persistent_term.get(cache_key, :not_found) do
```
I think rather than using something like persistent_term, I'd want to go through once at the end with a list of all the GitHub handles, and uniq them and fetch. That would make this logic generally simpler (`:persistent_term` AFAIK should only really be used for effectively "constants", not really as a cache).

TwistingTwists replied:
I agree that persistent_term is a hack. Given that this is a cli tool and not run frequently for a repo, this should be fine.

for `uniq` emails approach - would have to change more. Will try implementing that part

zachdaniel commented:
@TwistingTwists Any updates on this? Would be great to have this in ❤️

TwistingTwists commented:
Ah dang. I almost forgot about it. 

Will update this later this week

TwistingTwists committed:
fix: uniq emails per changelog and generate github username

- removed persistent_term
- use github_username if possilbe
- otherwise fallback to name from 'email' or commit.

TwistingTwists commented:
@zachdaniel ready for review. :)

zachdaniel commented:
🚀 Thank you for your contribution! 🚀
